### ISSUE TITLE
sysconfig: NM_CONTROLLED=False should not be set on RHEL8

### ISSUE THREAD

**Issue by ubuntu-server-builder** [canonical/cloud-init#3781]
This bug was originally filed in Launchpad as [LP: #1894837](https://bugs.launchpad.net/cloud-init/+bug/1894837)
<details>
<summary>Launchpad details</summary>
<pre>
affected_projects = []
assignee = None
assignee_name = None
date_closed = None
date_created = 2020-09-08T12:21:36.599977+00:00
date_fix_committed = None
date_fix_released = None
id = 1894837
importance = medium
is_complete = False
lp_url = https://bugs.launchpad.net/cloud-init/+bug/1894837
milestone = None
owner = amansi26
owner_name = Aman Kumar Sinha
private = False
status = triaged
submitter = amansi26
submitter_name = Aman Kumar Sinha
tags = []
duplicates = []
</pre>
</details>

_Launchpad user **Aman Kumar Sinha(amansi26)** wrote on 2020-09-08T12:21:36.599977+00:00_

Environment Details: 
Management Control Plane : OpenStack (Ussuri Release)
cloud-init version : 19.1 (community)
Data Source : Config Drive
OS/platform of deployed VM : RHEL 8.2

I am using cloud-init v19.1 where the control plane (OpenStack nova service) passes information (data source) via configdrive during VM deployment.

On a RHEL8.2 VM deployed from the above environment, it is observed that IPv4 interfaces do not come up. This behavior is observed only when NM_CONTROLLED is set to no in the interface files. This value is set from cloud-init src code at the below lines :

https://github.com/canonical/cloud-init/blob/stable-19.4/cloudinit/net/sysconfig.py#L275

we are setting NM_CONTROLLED = no using the code
    iface_defaults = tuple([
        ('ONBOOT', True),
        ('USERCTL', False),
        ('NM_CONTROLLED', False),
        ('BOOTPROTO', 'none'),
        ('STARTMODE', 'auto'),
    ])

under the file [1] . Due to which the Network Manager is not able to handle the interfaces

[1] cloudinit/net/sysconfig.py

When the above pieces of code is updated to set True to NM_CONTROLLED, the IPV4 interfaces come up fine.

Making ('NM_CONTROLLED', True) fixes the issue.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Chad Smith(chad.smith)** wrote on 2020-09-08T16:42:45.304727+00:00_

Please add cloud-init logs to this bug so we can triage a bit more. 

Logs can be obtained running `cloud-init collect-logs` on the vm and attaching the tar.gz to this bug.

Cloud-init specifically avoids trying to render network configuration for network-manager based interfaces as recommended in RHEL6 docs, but if RHEL8 needs to differ some documentation links around that would be helpful.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Chad Smith(chad.smith)** wrote on 2020-09-08T16:43:20.153531+00:00_

Feel free to mark this bug back to 'New' state at the top when you have a chance to respond so that we know it need review.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Aman Kumar Sinha(amansi26)** wrote on 2020-09-08T17:12:14.696156+00:00_

RHEL8 documentation:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/pdf/configuring_and_managing_networking/Red_Hat_Enterprise_Linux-8-Configuring_and_managing_networking-en-US.pdf
_Launchpad attachments_: [clod-init logs](https://api.launchpad.net/devel/cloud-init/+bug/1894837/+attachment/5408727/data)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Aman Kumar Sinha(amansi26)** wrote on 2020-09-09T16:39:00.986744+00:00_

IRC chat regarding this defect: https://irclogs.ubuntu.com/2020/09/08/%23cloud-init.html

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Aman Kumar Sinha(amansi26)** wrote on 2020-09-10T07:37:16.417381+00:00_

I validated this with cloud init provided by RHEL8.2 [cloud-init-19.4-1.el8.7.noarch] and I don't see NM_CONTROLLED  in the code under cloudinit/net/sysconfig.py

    iface_defaults = tuple([
        ('ONBOOT', True),
        ('USERCTL', False),
        ('BOOTPROTO', 'none'),
        ('STARTMODE', 'auto'),
    ])

Due to which the NM_CONTROLLED value does not get added in the interface file and by default NM_CONTROLLED value is yes.

Interface file looks like:

[root@rhle20 cloudinit]# cat /etc/sysconfig/network-scripts/ifcfg-env32
# Created by cloud-init on instance boot automatically, do not edit.
#
BOOTPROTO=none
DEFROUTE=yes
DEVICE=env32
GATEWAY=xxx.xxx.xxx.xxx
HWADDR=fa:c9:e6:d2:01:20
IPADDR=xxx.xxx.xxx.xxx
MTU=1500
NETMASK=xxx.xxx.xxx.xxx
ONBOOT=yes
STARTMODE=auto
TYPE=Ethernet
USERCTL=no

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Aman Kumar Sinha(amansi26)** wrote on 2020-09-10T08:15:57.557101+00:00_


_Launchpad attachments_: [RHEL8.2 provided  cloud-init v19.4](https://api.launchpad.net/devel/cloud-init/+bug/1894837/+attachment/5409305/data)

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Divya K Konoor(dikonoor)** wrote on 2020-09-10T08:16:49.359982+00:00_

So with cloud-init 19.4 version shipped as part of RHEL 8.2.1, we see a difference in code compared to what is found at https://github.com/canonical/cloud-init/blob/stable-19.4/cloudinit/net/sysconfig.py#L275. Due to this we are not able to reproduce this problem with RHEL cloud-init but can reproduce it with community cloud-init.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Divya K Konoor(dikonoor)** wrote on 2020-09-21T02:58:13.426842+00:00_

My understanding is that things work fine on RHEL7 with NM_CONTROLLED=False. So NM_CONTROLLED=True is a downstream patch maintained by RH only for RHEL8.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Paride Legovini(paride)** wrote on 2020-09-23T16:34:22.540381+00:00_

I retitled the bug to (hopefully) make it more to the point.

We should make sysconfig.py distinguish between RHEL < 8 and RHEL >= 8.

**Comment by ubuntu-server-builder** [comment]
_Launchpad user **Jessvin Thomas(jessvin)** wrote on 2021-11-16T14:49:42.843638+00:00_

I found one more item in the log. When I set the datasource to NoCloud I was able to see the following in the boot log:

[   11.319812] cloud-init[868]: 2021-11-16 14:47:50,680 - util.py[WARNING]: Running interface command ['nmcli', 'connection', 'up', 'ifname', 'eth0'] failed

So it looks like its trying to use nmcli but can't. 
Anywhere else I should look to troubleshoot?
